packages/gatsby/src/redux/nodes.ts

```diff
@@ -337,24 +393,109 @@ function addNodeToBucketWithElemMatch(
   }
 }
 
+const binarySearch = (
+  values: Array<FilterValue>,
+  needle: FilterValue
+): [number, number] => {
+  let min = 0
+  let max = values.length - 1
+  let pivot = Math.floor(values.length / 2)
+  while (min <= max) {
```

Comment:
can we end up in situation where `values.length === 0` -> `max = -1` -> `min <= max === false) -> reach ``// Shouldn't be reachable / return [0, 0]``?

Do `undefined`/`null` get added to indexes? If not then there can be scenario where field is defined by schema customization but there might no nodes with actual values there

Reply:
Yeah. But now that you mention it, it should return `undefined` to signal failure. Not an arbitrary node. I'll change that.

While adding `gte` I also identified a bug where `null` isn't being recorded at all (not a `FilterValue`. It also turns out that nodes for whom a path does not exist, are treated slightly different from nodes where it resolves to `null`. With `eq: null`, nodes with non-existing paths are included but `lte: null` and `gte: null` will only return nodes where the path explicitly resolves to null.

Since we don't use `undefined` (that's not a valid `FilterValue`) I'm going to put nodes which don't resolve for a given filter into a buck with `value=undefined`, to make the distinction against nodes that resolve explicitly to `null`. (Working this out right now)

Reply:
Ok, both of these have been addressed. Thanks! :)